aws/ec2_transit_gateway.go

```diff
@@ -0,0 +1,582 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func decodeEc2TransitGatewayRouteID(id string) (string, string, error) {
+	parts := strings.Split(id, "_")
+
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("Unexpected format of ID (%q), expected tgw-rtb-ID_DESTINATION", id)
+	}
+
+	return parts[0], parts[1], nil
+}
+
+func decodeEc2TransitGatewayRouteTableAssociationID(id string) (string, string, error) {
+	parts := strings.Split(id, "_")
+
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("Unexpected format of ID (%q), expected tgw-rtb-ID_tgw-attach-ID", id)
+	}
+
+	return parts[0], parts[1], nil
+}
+
+func decodeEc2TransitGatewayRouteTablePropagationID(id string) (string, string, error) {
+	parts := strings.Split(id, "_")
+
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("Unexpected format of ID (%q), expected tgw-rtb-ID_tgw-attach-ID", id)
+	}
+
+	return parts[0], parts[1], nil
+}
+
+func ec2DescribeTransitGateway(conn *ec2.EC2, transitGatewayID string) (*ec2.TransitGateway, error) {
+	input := &ec2.DescribeTransitGatewaysInput{
+		TransitGatewayIds: []*string{aws.String(transitGatewayID)},
+	}
+
+	log.Printf("[DEBUG] Reading EC2 Transit Gateway (%s): %s", transitGatewayID, input)
+	for {
+		output, err := conn.DescribeTransitGateways(input)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if output == nil || len(output.TransitGateways) == 0 {
+			return nil, nil
+		}
+
+		for _, transitGateway := range output.TransitGateways {
+			if transitGateway == nil {
+				continue
+			}
+
+			if aws.StringValue(transitGateway.TransitGatewayId) == transitGatewayID {
+				return transitGateway, nil
+			}
+		}
+
+		if aws.StringValue(output.NextToken) == "" {
+			break
+		}
+
+		input.NextToken = output.NextToken
+	}
+
+	return nil, nil
+}
+
+func ec2DescribeTransitGatewayRoute(conn *ec2.EC2, transitGatewayRouteTableID, destination string) (*ec2.TransitGatewayRoute, error) {
+	input := &ec2.SearchTransitGatewayRoutesInput{
+		// InvalidParameterValue: Value (transit-gateway-route-destination-cidr-block) for parameter Filters is invalid.
```

Comment:
Did you mean to leave these comments in here? Their purpose is not immediately clear to me (just a question, not a review item).

Reply:
Yeah, sorry I should've including a lead in note to those -- essentially its documenting that the EC2 documentation is incorrect by showing the errors. I'll push up an additional explanation there.